contracts/contracts/StealthKeyRegistry.sol

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.7;
+
+contract StealthKeyRegistry {
+  // =========================================== Events ============================================
+
+  /// @dev Event emitted when a user updates their resolver stealth keys
+  event StealthKeyChanged(
+    address indexed registrant,
+    uint256 spendingPubKeyPrefix,
+    uint256 spendingPubKey,
+    uint256 viewingPubKeyPrefix,
+    uint256 viewingPubKey
+  );
+
+  // ======================================= State variables =======================================
+
+  /// @dev The domain typehash used for EIP-712 signatures in setStealthKeysOnBehalf
+  bytes32 public constant EIP712DOMAIN_TYPEHASH =
+    keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)");
+
+  /// @dev The payload typehash used for EIP-712 signatures in setStealthKeysOnBehalf
+  bytes32 public constant STEALTHKEYS_TYPEHASH =
+    keccak256(
+      "StealthKeys(uint256 spendingPubKeyPrefix,uint256 spendingPubKey,uint256 viewingPubKeyPrefix,uint256 viewingPubKey)"
+    );
+
+  /// @dev The domain separator used for EIP-712 sigatures in setStealthKeysOnBehalf
+  bytes32 public immutable DOMAIN_SEPARATOR;
+
+  /**
+   * @dev Mapping used to store two secp256k1 curve public keys useful for
+   * receiving stealth payments. The mapping records two keys: a viewing
+   * key and a spending key, which can be set and read via the `setsStealthKeys`
+   * and `stealthKey` methods respectively.
+   *
+   * The mapping associates the user's address to another mapping, which itself maps
+   * the public key prefix to the actual key . This scheme is used to avoid using an
+   * extra storage slot for the public key prefix. For a given address, the mapping
+   * may contain a spending key at position 0 or 1, and a viewing key at position
+   * 2 or 3. See the setter/getter methods for details of how these map to prefixes.
+   *
+   * For more on secp256k1 public keys and prefixes generally, see:
+   * https://github.com/ethereumbook/ethereumbook/blob/develop/04keys-addresses.asciidoc#generating-a-public-key
+   *
+   */
+  mapping(address => mapping(uint256 => uint256)) keys;
+
+  /**
+   * @dev We wait until deployment to codify the domain separator because we need the
+   * chainId and the contract address
+   */
+  constructor() {
+    uint256 _chainId;
+    assembly {
+      _chainId := chainid()
+    }
+
+    DOMAIN_SEPARATOR = keccak256(
+      abi.encode(
+        EIP712DOMAIN_TYPEHASH,
+        keccak256(bytes("Umbra Stealth Key Registry")),
+        keccak256(bytes("1")),
+        _chainId,
+        address(this)
+      )
+    );
+  }
+
+  // ======================================= Set Keys ===============================================
+
+  /**
+   * Sets the stealth keys associated with an address, for anonymous sends.
+   * May only be called by the associated address.
```

Comment:
"anonymous sends" feels like a misleading description of Umbra. Maybe just describe what this method does instead of what the purpose of calling it is
```suggestion
   * @notice Sets stealth keys for the caller
```